### ISSUE TITLE
[Question] How to write Junit test for mapstruct abstract mapper injected via Spring

### ISSUE THREAD

**Issue by TehBakker** [mapstruct/mapstruct#1261]
I'm using MapStruct, mapstruct-jdk8 version 1.1.0.Final and defining abstract class that I inject via Spring.

I'm looking at how to be able to test them via Junit Test ?
I've basicaly a main mapper that will use 2 sub mappers

```
    @Mapper(componentModel = "spring", uses = {SubMapper1.class, SubMapper2.class})
    public abstract class MainMapper {
    
      @Mapping(target = "field1", qualifiedByName = {"MyMapper2Name", "toEntity"})
      public abstract MyEntity toEntity(MyDto pDto);
    
      public MyDto fromEntity(MyEntity pEntity) {
         // Specific code, hence why I use Abstract class instead of interface. 
      }
    }
```

I've tried several things but can't get the mapper to be instancied correctly to test it.

```
    @RunWith(SpringRunner.class)
    public class MainMapperTest {
    
        private MainMapper service = Mappers.getMapper(MainMapper.class);
    
        @Test
        public void testToEntity() throws Exception {
    .....
```

>  java.lang.RuntimeException: java.lang.ClassNotFoundException: Cannot find implementation for com.mappers.MainMapper

I've also tried via @InjectMock but no dice either.

> Cannot instantiate @InjectMocks field named 'service'. You haven't
> provided the instance at field declaration so I tried to construct the
> instance. However, I failed because: the type 'MainMapper is
> an abstract class.

And via Spring @Autowired 

> Caused by:
> org.springframework.beans.factory.NoSuchBeanDefinitionException: No
> qualifying bean of type 'com.mappers.MainMapper' available: expected
> at least 1 bean which qualifies as autowire candidate. Dependency
> annotations:
> {@org.springframework.beans.factory.annotation.Autowired(required=true)}

I'm guessing this might have to do with annotation processor, and mapper not being generated when I launch test.
I found this class as example : https://github.com/mapstruct/mapstruct/blob/master/processor/src/test/java/org/mapstruct/ap/test/decorator/spring/SpringDecoratorTest.java

However the class AnnotationProcessorTestRunner doesn't seems to be available before 1.2 which has no final release yet.

So my question is how do I write Junit tests to test my mapstruct abstract class mapper that I use via Spring injection in my code.
I don't want to call the impl constructor as I would like the sub mapper to be injected to test the whole process at once.

Sorry for the hassle.

**Comment by agudian** [comment]
The `AnnotationProcessorTestRunner` is our internal test runner used to test the annotation processor. It’s been there forever, but not part of the releases. And it’s not what you want anyway.

In short, you test your mappers with componentModel=“spring” just like any other spring bean. Using a component scan, a Spring `@Configuration` class or whatever you use in your project. The Impl is there during tests, as it has been generated while compiling your main classes.

**Comment by TehBakker** [comment]
Thank for the quick, answer.
Yes when I wire the spring context, I'll get my mapper, however then it'll try and wire all properties files and connect to the database, which is not what I want.
I tried to do only wire the mapper through @SpringBootTest(MyMapper.class) but then I get the instantiationException because it's abstract.

**Comment by TehBakker** [comment]
Thanks, got it working with a mix of @SpringBootTest @Autowired and @MockBean

**Comment by mohamnag** [comment]
And now another one has the same problem and should fight his way through. Given the fact that spring is a widely used framework, would be very helpful if a sample was provided to show how shall a 'spring' component model mapper be injected in tests. Thanks in advance.

**Comment by mohamnag** [comment]
sorry, just found this: https://github.com/mapstruct/mapstruct/blob/master/integrationtest/src/test/resources/springTest/src/test/java/org/mapstruct/itest/spring/SpringBasedMapperTest.java

**Comment by filiphr** [comment]
Great that you found that test. In a nutshell you can test the classes like you would any component that you have. In 1.3 there will be a constructor injection strategy and you would be able to instantiate the mappers by yourself more easily

**Comment by mohamnag** [comment]
thanks for the info, will be a better solution any way than setting up the whole test as an spring one to be necessary.

**Comment by rivancic** [comment]
**TL;DR**

I am using [ReflectionTestUtils#setField()](https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/test/util/ReflectionTestUtils.html#setField-java.lang.Object-java.lang.String-java.lang.Object-) to inject the concrete mapper to the field.

**Injecting field**

In case I need to test logical flow in the code without the need to use Spring Test Context, I inject few dependencies with Mockito framework. In case of MapStruct I need the concrete implementation that was generated from interface. The Mapper is used as a field. Here concrete class can be manually initialized with factory method and then set it to field with the help of ReflectionTestUtils.

Example test code:
```
@InjectMocks
private Service service;

@Before
public void setup() {
  MockitoAnnotations.initMocks(this);
  XMapper xMapper = Mappers.getMapper(XMapper.class); // Initialization of the mapper
  ReflectionTestUtils.setField(service, "fieldName", xMapper);
}
```

**Comment by chris922** [comment]
@rivancic this thread is already closed, but just one enhancement:

In case the mapper in your `Service` is annotated with `@Inject`, `@Autowired`, ... you can also let Mockito inject it with your `@InjectMocks` annotation.

You just need to define the mapper as a Mockito spy:

```java
@Spy
private XMapper xMapper = Mappers.getMapper(XMapper.class);

@InjectMocks
private Service service;
```

**Comment by rivancic** [comment]
@chris922, thanks for the hint. I did Autowired the mapper so the solution with `@Spy` annotation works perfectly.  :+1:

**Comment by filiphr** [comment]
One other thing that I can add to this is that with the upcoming 1.3.0 release there will be support for constructor injection. Which means that you won't need to use any special reflection to set the fields

**Comment by wagner8170** [comment]
Hi guys,

I've created this question regarding my approach to test mapper objects, can you check if is it ok?
https://stackoverflow.com/questions/57365149/junit-test-for-mapstruct-object?noredirect=1#comment101280160_57365149

thanks 😃

**Comment by vitorfranca-hotmart** [comment]
Im using InjectionStrategy.CONSTRUCTOR, componentModel spring, NOT using uses, using CycleAvoiding from examples and that worked for me. Doing tests normally, initializing the mappers like Mappers.getMapper(class).

**Comment by nipunarora** [comment]
> @rivancic this thread is already closed, but just one enhancement:
> 
> In case the mapper in your `Service` is annotated with `@Inject`, `@Autowired`, ... you can also let Mockito inject it with your `@InjectMocks` annotation.
> 
> You just need to define the mapper as a Mockito spy:
> 
> ```java
> @Spy
> private XMapper xMapper = Mappers.getMapper(XMapper.class);
> 
> @InjectMocks
> private Service service;
> ```

The Mapper Implementation uses some autowired beans, how can I mock these, since I see they are showing up as null?

**Comment by nrsina** [comment]
> > @rivancic this thread is already closed, but just one enhancement:
> > In case the mapper in your `Service` is annotated with `@Inject`, `@Autowired`, ... you can also let Mockito inject it with your `@InjectMocks` annotation.
> > You just need to define the mapper as a Mockito spy:
> > ```java
> > @Spy
> > private XMapper xMapper = Mappers.getMapper(XMapper.class);
> > 
> > @InjectMocks
> > private Service service;
> > ```
> 
> The Mapper Implementation uses some autowired beans, how can I mock these, since I see they are showing up as null?

I'm having the same issue. 
My "PersonMapper" abstract class have some dependencies on other beans..
I would be most grateful if someone could give me advice regarding this issue.
Kind Regards.

**Comment by mgrao1** [comment]
Good morning,

The same thing happened to me.

I had an abstract mapper that used 1 bean. In order to instantiate it and make it work I did the following.

@Spy
private XMapper xMapper = Mappers.getMapper(XMapper.class);

@Mock
private ServiceImpl service;

@Before
public void init() {
ReflectionTestUtils.setField(xMapper , "service", service);
}

I hope you find it useful
Regards.

**Comment by mgrao1** [comment]
Good morning,

The same thing happened to me.

I had an abstract mapper that used 1 bean. In order to instantiate it and make it work I did the following.

`@Spy`
`private` XMapper xMapper = `Mappers.getMapper(XMapper.class);`

`@Mock`
`private serviceImpl service;`

`@Before`
`public void init() {`
`ReflectionTestUtils.setField(xMapper , "service", service);`
`}`


I hope you find it useful
Regards.

**Comment by pumano** [comment]
micro lifehack for people who want to write tests for mappers:

Just add `unmappedTargetPolicy = ERROR` and application fail to build if some fields can't be mapped, also same for source if needed. It's pretty easy unit-test without zero testing code.

**Comment by dalvan-bevilaqua** [comment]
> 
thank you so much

**Comment by dalvan-bevilaqua** [comment]
to me the solution @chris922
